marge/single_merge_job.py

```diff
@@ -49,46 +51,57 @@ def update_merge_request_and_accept(self, approvals):
 
         while not updated_into_up_to_date_target_branch:
             self.ensure_mergeable_mr(merge_request)
-            source_project, source_repo_url, _ = self.fetch_source_project(merge_request)
-            target_project = self.get_target_project(merge_request)
-            try:
-                # NB. this will be a no-op if there is nothing to update/rewrite
 
-                target_sha, _updated_sha, actual_sha = self.update_from_target_branch_and_push(
-                    merge_request,
-                    source_repo_url=source_repo_url,
+            if self._config.use_only_gitlab_api and self._options.fusion is Fusion.gitlab_rebase:
+                source_branch = Branch.fetch_by_name(
+                    merge_request.source_project_id,
+                    merge_request.source_branch,
+                    self._api,
                 )
-            except GitLabRebaseResultMismatch as err:
-                if first_iteration:
-                    log.info(
-                        "Gitlab rebase didn't give expected result."
-                        "This is expected immediately after rebase. Retrying."
+
+                if source_branch.commit.id != merge_request.diff_refs.base_sha:
+                    self.synchronize_using_gitlab_rebase(merge_request)
```

Comment:
Here's the most important fix. Instead of always going through the `self.fetch_source_project`, `self.get_target_project`, `self.update_from_target_branch_and_push` code paths, we just check if the SHA of the target branch HEAD and the base SHA of the MR matches, otherwise we assume the MR needs to be rebased. The `self.update_from_target_branch_and_push` call requires a local git repo checkout.

This reminds me that we should probably do a final check that the target branch HEAD SHA and MR base SHA matches after CI has passed too.

Reply:
Ah, but that is not needed :thinking: It's already handled.

![image](https://user-images.githubusercontent.com/1417619/211284172-e879d410-304f-4a22-a8eb-829af15d6a2d.png)